### DISCUSSION TITLE
How to fully benefit from server rendering; organizing loader & action functions?

### DISCUSSION THREAD

**Discussion by jdevries3133** [remix-run/remix#2599] [Q&A]
Hi All,

I've been experimenting a bunch with [remix.run](https://remix.run). I am enjoying the framework overall, but I'm really struggling with how to benefit the most from the client-side rendering model. Something I liked about remix on first learning about it was how heavily it relied on web fundamentals. I thought I could write a highly dynamic app that worked just as well for clients not using javascript, while giving up very little in terms of my own developer experience.

To achieve this with remix, you need to keep all the data mutation inside `loader` and `action` functions, and these functions must be exported only by route modules. This is all fine and good, but it also forces me to have just one pair of functions to deal with all the data mutations for a whole page (or at least a big chunk of it). Of course, remix does have nested routes, but they are just that – nested. It works well if you have a hierarchy, like `sitewide header > sidebar > section header > object picker > object`, which is kind of the situation they show in the demo on their homepage. However, not every page is laid out hierarchically. Instead, there might be a sequence of multiple components that are all equal in hierarchy. Suddenly, I find myself doing a pattern like this:

```javascript
// /routes/dashboard/panelOne.jsx

// ... loaders and actions for panelOne live here

export default function PanelOne() {
  return <><div>panel 1...</div><Outlet /></>;
}
```

```javascript
// /routes/dashboard/panelOne/panelTwo.jsx

... loaders and actions

export default function PanelTwo() {
  return <><div>Panel 2</div><Outlet /></>
```

```javascript
// /routes/dashboard/panelOne/panelTwo/panelThree.jsx

// ... loaders & actions

export default function PanelThree() {
  return <><div>panel 3...</div><Outlet /></>
}
```

In the end, this is the page we want:

- panel 1
- panel 2
- panel 3

But, we have to structure it hierarchically so that each panel can get its own loader and action function

- panel 1
  - panel 2
    - panel 3

The URL is similarly weird. Just to see the dashboard, you have to go to:

`/dashboard/panelOne/panelTwo/panelThree`

That hierarchy doesn't make any sense!

Now, I guess that one possible solution to this conundrum is just to forget about action and loader functions, let those deeply nested components manage their own state or do their own data fetching, use a state management library, etc. Basically, "do things the old fashioned way" from a React point of view. However, I feel like that misses the whole point of the framework and leads me towards not supporting non-js clients very well.

The further I unwrap this whole situation, I feel like I only end up with more questions, like:

- Why can't I just put a loader or action into any arbitrary component? I feel like that would be a ridiculously fantastic improvement to the framework? I know that it creates a problem i.e., "well, what does the URL for that data endpoint become?" but it's not a user-facing URL. I feel like it can just be a random hash.
- I feel like my route modules always end up super bloated. It *can* be clean if only one thing is happening on that route, but as soon as you have a page with a few different modules, each of which need their own data, you get into hot water immediately.
- Am I overthinking the hierarchy thing? Should I just make a spider's web of route files like a madman? I feel like route files are a bit messy when they actually hold anything, and it's better to factor most of the logic and view into services and components, respectively. Is this wrong?
- Am I making my route files too big? I know that remix has subroutes; will I benefit from just taking this idea a lot farther? I feel like I am going to end up with comically long URLs if I go down this road. Either that, or atrocious filenames because I need to namespaces things like `/b/d/c/`, and my files then become `b.jsx` (*shudders*)

**Comment by kiliman** [comment]
Remix loads data at route boundaries via `loader`. If you need data in your components that can't be returned from your loader, then you can always create a Resource route and `useFetch` from the component to get the data.

As for actions, you can have a single action process multiple types of actions. Simply include a name/value pair that you use to distinguish. For example, `action=toggle`, `action=delete`, etc.
```ts
export const action: ActionFunction = async ({request}) => {
  const formData = await request.formData()
  switch (formData.get('action')) {
    case 'toggle': 
      doToggle(formData)
      break
    case 'delete': 
      doDelete(formData)
      break
    ...
  }
}
```

NOTE: As for big loaders/actions, you don't need to keep all your logic in the route module. You can import functions that do the work. The only requirement is that you export a function named `loader` or `action`. The actual code can reside in separate files.

Checkout Remix Singles by Ryan Florence where he talks about loaders and actions

https://www.youtube.com/watch?v=Iv25HAHaFDs
https://www.youtube.com/watch?v=w2i-9cYxSdc

If you really want to manage data at the component level, check out my POC where I show you how to export loaders/actions from components.
https://github.com/kiliman/remix-component-data

**Reply by jdevries3133** [reply]
This is super helpful, particularly looking at that demo app you created. Thank you!

**Reply by sgoodrow** [reply]
Thanks for the solution. It works, but definitely feels like a workaround for a fundamental gap in the framework. The amount of boilerplate to connect components to their data feels awkward. 

@kentcdodds is a large advocate for DX, I'm curious what he thinks. I think the Full Stack Components approach he shared is heading in this direction, but still missing some gaps (error boundaries) and leans more on defining new patterns than reusing existing ones.

**Comment by sergiodxa** [comment]
What I would do in your case, is to create those components outside the route folder and export a function together with the component to get the data they need, also export the types.

Then, in the route importing and rendering the components I would also import those functions and call them in the loader, all in parallel.

**Reply by sgoodrow** [reply]
Apologies for being so late to the party. The issue with this approach is that we don't have access to the error boundary pattern for each component, right?

**Comment by sgoodrow** [comment]
I think `useFetcher` and this [Full Stack Components ](https://www.epicweb.dev/full-stack-components) pattern are close (and an affirmation of the problem). My only holdup is that they don't expose an easy remix-like way to add error boundaries to the full stack components, **and that it's a bit of API bloat.**

Dreaming: I'd love if full stack components were first class citizens of remix, where the same patterns for defining routes can apply to defining components, and their component filepath is used as the hashing function (e.g. we can define a `components` directory that behaves just like `routes`, except they are imported and instantiated into routes).

One of the biggest successes of React 16 in my eyes is the very narrow API that serves most of the typical webdev use cases. We should strive for something similar in Remix, as this increases accessibility and minimizes developer cognitive load, making us all more productive developers.

**Reply by kiliman** [reply]
I have a POC for component-level loaders/actions and Error/Catch boundaries. You still have to wire stuff up manually though, but could probably write something to code-gen that.

https://github.com/kiliman/remix-component-data

I need to move sample app to another host since Heroku no longer provides a free tier.

**Comment by sgoodrow** [comment]
I think this is the right direction, but yeah the wiring is unfortunately clumsy imo. Definitely better than nothing and a good poc!

**Comment by runarj** [comment]
"Remix Component Data" by @kiliman seems promising, but I wish the problems were addressed in the _core_ of Remix instead. I like the ideas in Remix, but I am also confused about how to structure an application that shows a lot of data.

React Router (which is used by Remix) defines the rendered components. The components created are determined from the URL, which – in reality – is a **path down a hierarchy.** To display the mentioned dashboard with the panels, you do not want to render a path down a hierarchy but a **tree** of depth 2. But representing an entire tree/hierarchy in an URL is tricker than representing a path.

When you want to display multiple documents and multiple detail pages at the _same time_, maybe the way components are created by React Router will get in your way? Perhaps trying to represent the structure of components using React Router fits better for a "drill-down"-centered UI, than for a UI that displays an entire tree of objects?

But if one switches from letting React Router manage components and starts using something like MobX "UI stores," there is nothing of Remix left. Remix uses React Router to define the components, and the component is where all the fun happens.

An URL is a string that most often represents a path down a hierarchy, but it is possible to let the string define a tree instead. But as far as I understand, React Router is not ready to represent a tree of components. (This might be a proposal to change that, but I do not understand the details and implications: https://github.com/remix-run/react-router/discussions/8023 )

Should React Router be enhanced to support a UI created as a tree, not a path down a tree?

Should Remix support methods other than React Router to define the components?

I am new to Remix, so maybe I misunderstand the whole concept. Please correct my thoughts!

**Reply by jdevries3133** [reply]
As I've been using remix for apparently 11 months now (geez the time flies), I feel like I've mostly solved this problem for myself, but I think the remix DX can go further to make the solution more obvious. Clearly, Killimans solution of factoring logic out of actions and loaders does a good job:

```typescript
export async const loader = () => ({
  posts: await loadPosts(),
  comments: await loadComments()
})
```

What's not-so-obvious is how to hand this with typescript. I've been doing it like this:

```typescript
type PostLoaderData = {
  posts: Post[]
}

type CommentLoaderData = {
  comments: Comment[]
}

// Then, in the frontend
const { posts, comments } = useLoaderData<PostLoader & CommentLoader>();
```

This is nice because I can define the PostLoaderData type in the module with the loadPosts function. It's also nice because I can usually define this type using the loader itself, so everything stays in sync:

```typescript
type PostLoaderData = Awaited<ReturnType<typeof loadPosts>>
```

It's also nice because I have composability. If one component only depends on the comments in the loader, I can define the type more narrowly, which usually is the case. 

A big annoyance here is that typescript can't guarantee that the component is always imported into pages where it will get the loader data it expects. This is the #1 issue I have with this approach but I'm pretty sure that the limitations of typescript with file based routing make this impossible to solve; at least I have no clue how I would approach the solution. 

In the meantime, some API to help people with the plumbing and some documentation improvement would go a long way, I think. 

I also think that there's something "right" about remix's page-level-loader abstraction. Notably, it forces you to funnel all of a page's data through a single point in your application. That's great! It feels like a good practice, and the reality is that if you want a performance page, you have to fetch with care. It makes sense for the developer to be faced with this loader that does a big job, because it is a single point where you can understand all the data manipulation behavior of a page. 

On the other hand, I think that the single action function is probably more annoying. I want more control over the granularity of data manipulation. After page load, actions on a rich page happen slowly over time, and a single action will often have a lot of code attached to it. I'd really prefer for every form (even teeny tiny ones) to have their own server-side handler in the same file, which is loosely coupled with everything else.

**Reply by runarj** [reply]
I partially understand your reflections, @jdevries3133. But let us say we have an application that starts with two (or more) independent components at the top level. Then there will be a single top-level route and no React Router sub-routes.

If the two sub-components are of equal importance, it might be hard to decide which should have additional sub-routes. One example can be a system where you can simultaneously edit a set of documents, each taking up a part of your (visible) screen.

Would your chosen approach make sense when a large part of the application is _nested components_ and not routes?
